cmake: add ExternalProjects to enable QAT by default

cbodley opened the pull request:
fedora/centos/rhel now have system packages for `qatlib-devel` and `qatzip-devel`

adds submodules for qatlib/qatzip, along with `BuildQAT.cmake` and `BuildQATzip.cmake` to build these libraries as `ExternalProject`s. we'll rely on these submodules for ubuntu/debian until the deb packages show up

~~Ceph includes a `qae_mem_utils.h` header in [src/crypto/qat/qcccrypto.h](https://github.com/ceph/ceph/blob/07614d51c26cb3e98c87f4d9984a033b1e11761d/src/crypto/qat/qcccrypto.h#L30) that isn't present in the `qatlib-devel` package or the 'make install' directory of a qatlib build from source. i've opened https://github.com/intel/qatlib/pull/64 to address this upstream, and added a workaround to `BuildQAT.cmake` that manually copies the header into its install dir. until this is resolved and backported, we won't be able to rely on the `qatlib-devel` package~~

on the other hand, the `qatzip-devel` package works great and `ceph.spec.in` enables `WITH_SYSTEM_QATZIP` by default

renamed `WITH_QAT` to `WITH_QATDRV` for the out-of-tree driver, and added aliases to the `QAT::qat` and `QAT::usdm` targets from `FindQAT.cmake`

<details>
<summary>Show available Jenkins commands</summary>

- `jenkins retest this please`
- `jenkins test classic perf`
- `jenkins test crimson perf`
- `jenkins test signed`
- `jenkins test make check`
- `jenkins test make check arm64`
- `jenkins test submodules`
- `jenkins test dashboard`
- `jenkins test dashboard cephadm`
- `jenkins test api`
- `jenkins test docs`
- `jenkins render docs`
- `jenkins test ceph-volume all`
- `jenkins test ceph-volume tox`
- `jenkins test windows`
- `jenkins test rook e2e`
</details>

cbodley commented:
cc @hualongfeng @liumiaomiaoIntel @majianpeng

cbodley commented on cmake/modules/BuildQAT.cmake:
```diff
@@ -0,0 +1,55 @@
+  # 'make install' is missing one header that ceph requires, so copy it manually
+  list(APPEND install_cmd COMMAND cmake -E copy -t ${QAT_INCLUDE_DIR}/qat ${QAT_SOURCE_DIR}/quickassist/utilities/libusdm_drv/include/qae_mem_utils.h)
+
+  set(source_dir_args
+    SOURCE_DIR ${QAT_SOURCE_DIR}
+    GIT_REPOSITORY ${QAT_REPO}
+    GIT_TAG ${QAT_TAG}
+    GIT_SHALLOW TRUE
+    GIT_CONFIG advice.detachedHead=false)
```
if we're enabling these by default, it probably makes more sense to add the repos as submodules and point the external projects at their source dirs

liumiaomiaoIntel commented on cmake/modules/FindQAT.cmake:
```diff
@@ -0,0 +1,40 @@
+  HINTS ${PC_QAT_INCLUDE_DIRS})
+
+find_library(QAT_LIBRARY
+  NAMES qat qat_s
```
Here is the thing, if users' system isn't rhel>9 and their QAT devices are earlier than gen 4 devices, they need to install [out-of-tree QAT driver)](https://www.intel.com/content/www/us/en/download/19734/intel-quickassist-technology-driver-for-linux-hw-version-1-x.html). Becasue QAT gen 4 devices support qatlib, and earlier generations of QAT devices (e.g. c62x, dh895xxcc, etc.) are not supported by in-tree qat driver and [qatlib](https://github.com/intel/qatlib). By installing  out-of-tree QAT driver, there is no `qat/cpa.h` in system.

cbodley replied:
i see, thanks. so the `FindQatDrv.cmake` file i replaced was specific to the out-of-tree build

the qatlib version is probably(?) the only one that's reasonable to package/ship to users by default, but we can preserve the ability to build with `FindQatDrv.cmake`. for example, `WITH_QAT` would build against qatlib by default, but you could select the other with `WITH_QATDRV` and crypto/qat/qcccrypto.h would decide which headers to include based on `#ifdef HAVE_QATDRV`

does that sound ok?

liumiaomiaoIntel replied:
great. Coincidentally, I have also been adding the function to build qatlib and QATzip automatically when compile Ceph in the past week and I'm almost done with the test. The main differences are: 1.  `WITH_QAT=ON` would find out-of-tree qat driver in system, which is basically unchanged ; `WITH_QATLIB` would build qatlib by Ceph compilation, which is basically the same as your `BuildQAT.cmake`. But I also **added a scene to install qatlib in [standalone mode](https://github.com/intel/qatlib/blob/abe15d7bfc083117bfbb4baee0b49ffcd1c03c5c/INSTALL#L496C5-L496C5) which is essential for Ceph to run on containers** 2. `WITH_QATZIP=ON` would first find out whether there is an installed QATzip library in the system, and if not, it will pull [QATzip code](https://github.com/intel/QATzip) and install it, which  is similar with your `BuildQATzip.cmake`, but distinguished between _out-of-tree qat driver_ and qatlib.
Your basic framework is more complete than mine. But IMHO, functionally, support for out-of-tree qat drver(i.e. support for earlier generations of QAT devices, e.g. c62x, dh895xxcc, etc.) cannot be deleted.

cbodley replied:
that is a happy coincidence :)

how would you like to collaborate on this? you're welcome to take any of these changes into a branch/pr of your own. or i can adapt what i have here to use WITH_QAT/WITH_QATLIB/QAT_QATZIP as you've described, and we build on that

liumiaomiaoIntel replied:
Thanks so much. Your basic framework is very complete. I would like to make some changes based on your branch, mainly including installation compatibility of out-of-tree QAT dirver and the enablement of containerized ceph scene.

liumiaomiaoIntel commented on src/crypto/qat/qcccrypto.h:
```diff
@@ -16,18 +16,18 @@
+#include <qat/qae_mem.h>
+#include <qat/qae_mem_utils.h>
+#include <qat/icp_sal_user.h>
+#include <qat/icp_sal_poll.h>
```
maybe error for earlier generations of QAT devices (e.g. c62x, dh895xxcc, etc.)

liumiaomiaoIntel commented on cmake/modules/BuildQAT.cmake:
```diff
@@ -0,0 +1,55 @@
+function(build_qat)
+  set(QAT_REPO https://github.com/intel/qatlib.git)
```
earlier generations of QAT devices (e.g. c62x, dh895xxcc, etc.) are not supported by in-tree qat driver and [qatlib](https://github.com/intel/qatlib).

liumiaomiaoIntel commented on cmake/modules/BuildQATzip.cmake:
```diff
@@ -0,0 +1,46 @@
+  # imported targets need it early for INTERFACE_INCLUDE_DIRECTORIES
+  file(MAKE_DIRECTORY "${QATzip_INCLUDE_DIR}")
+
+  set(configure_cmd env CC=${CMAKE_C_COMPILER} ./configure --prefix=${QATzip_INSTALL_DIR})
```
earlier generations of QAT devices (e.g. c62x, dh895xxcc, etc.) are not supported by in-tree qat driver and [qatlib](https://github.com/intel/qatlib), and they only support out-of-tree qat driver. [QATzip](https://github.com/intel/QATzip) installation is different with qatlib and out-of-tree qat driver.

liumiaomiaoIntel commented on CMakeLists.txt:
```diff
@@ -325,9 +325,18 @@ endif()
 if(WITH_QAT)
+  if(NOT WITH_SYSTEM_QAT)
+    include(BuildQAT)
+    build_qat()
```
earlier generations of QAT devices (e.g. c62x, dh895xxcc, etc.) are not supported by in-tree qat driver and [qatlib](https://github.com/intel/qatlib). and user need to install out-of-tree qat driver. in this way, there maybe not proper to install qat by ceph.

cbodley commented:
changes:
* renamed WITH_QAT to WITH_QATLIB
* preserved FindQatDrv.cmake with renamed option WITH_QATDRV. adds ALIAS targets to match targets from FindQAT.cmake
* BuildQATzip.cmake supports WITH_QATDRV build using `configure --enable-ICP_ROOT=`
* qcccrypto.h includes depend on `#ifdef HAVE_QATDRV`

cbodley commented on src/CMakeLists.txt:
```diff
@@ -507,7 +507,8 @@ if(NOT WITH_SYSTEM_BOOST)
 
 if(HAVE_QATZIP)
-  list(APPEND ceph_common_deps ${qatzip_LIBRARIES})
+  # TODO: only the compression plugins should depend on QAT
+  list(APPEND ceph_common_deps QAT::qat QAT::usdm QAT::zip)
```
i did this as a separate https://github.com/ceph/ceph/pull/54653 since it doesn't depend on anything here

mkogan1 commented:
@cbodley just note/fyi for future reference, trying to build on RHEL9 _without_ `-DWITH_SYSTEM_QATLIB=ON -DWITH_SYSTEM_QATZIP=ON` & clang 16, resulted in qatzip build errors:
```
Making all in src                                                                                                                                                                            
gmake[1]: Entering directory '/mnt/raid0/src-git/ceph--mk--wip/src/qatzip/src'                                                                                                               
  CC       libqatzip_la-qatzip.lo                                                                                                                                                            
  CC       libqatzip_la-qatzip_counter.lo                                                                                                                                                    
  CC       libqatzip_la-qatzip_gzip.lo                                                                                                                                                       
In file included from qatzip_counter.c:51:                                                                                                                                                   
./qatzip_internal.h:445:25: error: a function declaration without a prototype is deprecated in all versions of C [-Werror,-Wstrict-prototypes]                                               
void streamBufferCleanup();                                                                                                                                                                  
                        ^                                                                                                                                                                    
                         void                                                                                                                                                                
1 error generated.         
```

mkogan1 commented on CMakeLists.txt:
```diff
@@ -325,16 +325,43 @@ endif()
 
-option(WITH_QAT "Enable Qat driver" OFF)
-if(WITH_QAT)
+option(WITH_QATLIB "Enable QAT with qatlib" ON)
```
@cbodley pushed to CI to try a Shaman build:
https://shaman.ceph.com/builds/ceph/wip-mk-up-QAT-i001/
there are QAT build errors on _Arm arch_ and _Windows  distro_ since default of `WITH_QATLIB=ON`
but myb there should be an exception for those and set them to `WITH_QATLIB=OFF` by default?

cbodley replied:
> the arm build is failing to build qatlib samples with `error: impossible constraint in 'asm'` - i'll try adding `--disable-samples` to the configure line

[arm64 build](https://jenkins.ceph.com/job/ceph-dev-new-build/ARCH=arm64,AVAILABLE_ARCH=arm64,AVAILABLE_DIST=centos8,DIST=centos8,MACHINE_SIZE=gigantic/74425//consoleFull) still failed on:
```
quickassist/lookaside/access_layer/src/qat_direct/common/adf_user_ring.c: In function ‘adf_user_put_msg’:
./quickassist/lookaside/access_layer/src/qat_direct/common/include/adf_platform_common.h:231:9: error: unknown register name ‘%edi’ in ‘asm’
  231 |         __asm__ __volatile__("mov %0, %%esi \n"                                \
      |         ^~~~~~~
```

liumiaomiaoIntel commented on cmake/modules/BuildQAT.cmake:
```diff
@@ -0,0 +1,48 @@
+
+  set(configure_cmd env CC=${CMAKE_C_COMPILER} ./configure --prefix=${QAT_INSTALL_DIR})
+  # disable systemd or 'make install' tries to write /usr/lib/systemd/system/qat.service
+  list(APPEND configure_cmd --disable-systemd)
```
The option `--disable-systemd` will install qatlib in [standalone mode](https://github.com/intel/qatlib/blob/abe15d7bfc083117bfbb4baee0b49ffcd1c03c5c/INSTALL#L257), so there is no _qat_ service running, which [is the preferred way to run on containers where a subset of VFs are explicitly passed through and so no qatmgr managed resource mgr is needed.](https://github.com/intel/qatlib/blob/abe15d7bfc083117bfbb4baee0b49ffcd1c03c5c/INSTALL#L499C8-L499C8)  However IMHO, if the Ceph cluster is running on bare metal not in container, it maybe better to install qatlib in [managed mode](https://github.com/intel/qatlib/blob/abe15d7bfc083117bfbb4baee0b49ffcd1c03c5c/INSTALL#L366) with option `--enable-service` so that qatmgr manages allocation of the VF resources to each process on process startup from performance perspective. On the other hand, if qatlib packages is installed on Linux distributions such as Fedora 34+, RHEL 9.1+, or CentOS 9 Stream by running command `sudo dnf -y install qatlib-devel`, I think [qatlib is installed in managed mode](https://intel.github.io/quickassist/qatlib/install.html#installing-from-rpm-packages), where _qat_ service is enabled. IMHO, it maybe better to distinguish the installation between managed mode and standalone mode. Although it seems to be able to run mutil-process test to use QAT devices by installing qatlib in standalone mode on bare metal, the performance of qat is worse than its installation in managed mode.

cbodley replied:
oops, thanks. i'd only added --disable-systemd because the 'make install' step doesn't respect the --prefix, and we don't want to be installing things to system directories (or require root) during the ceph build. we could work around that by avoiding the 'make install' step and pointing at artifacts in the source/build directories instead of the install dir

but we'd also need to package the qat service and its systemd stuff to be installed along with ceph, and that gets complicated and risks conflicts if someone installs ceph on a system that also has qat packages installed

it's also tricky to do it one way (standlone) for container builds and another way (managed) for package builds, because afaik our container images are built from the ceph packages

it does make sense for cmake to offer both options - and maybe WITH_SYSTEM_QATLIB=ON means managed and OFF means standalone - but we may not be able to support both for upstream users

our hands are further tied by the fact that debian doesn't have system packages, so the WITH_SYSTEM_QATLIB=OFF build would be the only thing we could ship atm

cbodley replied:
@liumiaomiaoIntel @mkogan1 @mattbenjamin is everyone okay with this decision to only support the standalone mode for ceph releases (both packages and containers)?

> In standalone mode, there is no qat service running. This is the preferred way to run on containers where a subset of VFs are explicitly passed through and so no qatmgr managed resource mgr is needed.

on the container side, do we need any additional support from cephadm to accomplish this part?

liumiaomiaoIntel replied:
Intel QAT devices delivers turn-key standard PCI Express access to hardware acceleration and their resources need to be managed to allocate QAT de/compression and cryption VFs to user's processes. In bare metal, qat service installed by the managed mode is used to manage QAT resources. In container, QAT VF resources also need to be managed, for example [QAT device plugin](https://github.com/intel/intel-device-plugins-for-kubernetes/blob/main/cmd/qat_plugin/README.md) who manages QAT resources in Kubernetes. In this way, cephadm might need to manage QAT VFs and path through a subset of VFs to each OSD containers without conflict.

cbodley replied:
@liumiaomiaoIntel that's very helpful, thank you

cbodley commented:
all of the [shaman builds](https://shaman.ceph.com/builds/ceph/wip-cmake-qat/96d1aa1b8982a07a1b3509d2ab542783298e510e/) are struggling with qatlib:
```
[  7%] Performing build step for 'qatlib_ext'
CMake Error at /home/jenkins-build/build/workspace/ceph-dev-new-build/ARCH/x86_64/AVAILABLE_ARCH/x86_64/AVAILABLE_DIST/centos8/DIST/centos8/MACHINE_SIZE/gigantic/release/19.0.0-45-g96d1aa1b/rpm/el8/BUILD/ceph-19.0.0-45-g96d1aa1b/x86_64-redhat-linux-gnu/qatlib_ext-prefix/src/qatlib_ext-stamp/qatlib_ext-build-.cmake:37 (message):
  Command failed: 1

   '/usr/bin/cmake' '-E' 'env' '--unset=DESTDIR' '$(MAKE)' '-j3'

  See also

    /home/jenkins-build/build/workspace/ceph-dev-new-build/ARCH/x86_64/AVAILABLE_ARCH/x86_64/AVAILABLE_DIST/centos8/DIST/centos8/MACHINE_SIZE/gigantic/release/19.0.0-45-g96d1aa1b/rpm/el8/BUILD/ceph-19.0.0-45-g96d1aa1b/x86_64-redhat-linux-gnu/qatlib_ext-prefix/src/qatlib_ext-stamp/qatlib_ext-build.log

-- Log output is:
No such file or directory
```

mattbenjamin commented on src/crypto/qat/qcccrypto.cc:
```diff
@@ -66,7 +66,7 @@ auto QccCrypto::async_get_instance(CompletionToken&& token) {
       // keep a few objects to wait QAT instance to make sure qat full utilization as much as possible,
       // that is, QAT don't need to wait for new objects to ensure
       // that QAT will not be in a free state as much as possible
-      instance_completions.push_back([this, ex, handler2 = std::move(handler1)](int inst)mutable{
```
does removing "this" from the lambda capture have any discrete purpose in the PR?

cbodley replied:
the first 3 commits resolve compile warnings/errors. i should have documented those in the commit messages. this one only happens on clang:
```
[17/19] Building CXX object src/crypto/qat/CMakeFiles/ceph_crypto_qat.dir/qcccrypto.cc.o
src/crypto/qat/qcccrypto.cc:68:39: warning: lambda capture 'this' is not used [-Wunused-lambda-capture]
   68 |       instance_completions.push_back([this, ex, handler2 = std::move(handler1)](int inst)mutable{
      |                                       ^
```

mattbenjamin commented on src/crypto/qat/qcccrypto.cc:
```diff
@@ -39,7 +39,6 @@ static void symDpCallback(CpaCySymDpOpData *pOpData,
   }
 }
 
-static std::mutex qcc_alloc_mutex;
```
I'd be interested in knowing why the author(s) thought we might need qcc_alloc_mutex

cbodley replied:
use of that mutex was removed in https://github.com/ceph/ceph/pull/47040 because it could guarantee that all access happened on the same background thread

mattbenjamin commented on src/compressor/CMakeLists.txt:
```diff
@@ -1,19 +1,15 @@
-if (HAVE_QATZIP)
-  list(APPEND compressor_srcs QatAccel.cc)
-endif()
-add_library(compressor_objs OBJECT ${compressor_srcs})
```
I do think this was a good change

tchaikov commented on src/crypto/qat/qcccrypto.cc:
```diff
@@ -539,7 +539,7 @@ bool QccCrypto::symPerformOp(int avail_inst,
       poll_retry_num = RETRY_MAX_NUM;
       if (y) {
-        yield_context yield = y.get_yield_context();
+        spawn::yield_context yield = y.get_yield_context();
```
probably might want to move the first three change into a separate PR. they can be merged much sooner than the CMake ones. i believe.

cbodley replied:
i'm willing to split stuff off if the cmake part gets blocked, but there are benefits to keeping them together for now - the cmake changes depend on them, and also allow us to verify that they build in CI

tchaikov commented on CMakeLists.txt:
```diff
@@ -314,10 +314,10 @@ endif()
 option(WITH_SYSTEM_QATLIB "Use system packages for qatlib" OFF)
 option(WITH_QATDRV "Enable QAT with out-of-tree driver" OFF)
-option(WITH_QATZIP "Enable QATzip" OFF)
+option(WITH_QATZIP "Enable QATzip" ON)
```
does this deserve a line in the release note? i am not sure about how important this is though.

cbodley replied:
good idea, added a release note

tchaikov commented on src/crypto/qat/qcccrypto.h:
```diff
@@ -28,6 +30,16 @@ extern "C" {
 #include "icp_sal_poll.h"
 #include "qae_mem_utils.h"
+#else
+#include <qat/cpa.h>
```
just realized that QATDrv installs the headers into a quite different path -- `quickassist/include` =(

mkogan1 commented:
> all of the [shaman builds](https://shaman.ceph.com/builds/ceph/wip-cmake-qat/96d1aa1b8982a07a1b3509d2ab542783298e510e/) are struggling with qatlib:
> 
> ```
> [  7%] Performing build step for 'qatlib_ext'
> CMake Error at /home/jenkins-build/build/workspace/ceph-dev-new-build/ARCH/x86_64/AVAILABLE_ARCH/x86_64/AVAILABLE_DIST/centos8/DIST/centos8/MACHINE_SIZE/gigantic/release/19.0.0-45-g96d1aa1b/rpm/el8/BUILD/ceph-19.0.0-45-g96d1aa1b/x86_64-redhat-linux-gnu/qatlib_ext-prefix/src/qatlib_ext-stamp/qatlib_ext-build-.cmake:37 (message):
>   Command failed: 1

noticed that there isent anything between the `qatlib_ext-build-` and the `.cmake`
like in the ... `qatlib_ext-prefix/src/qatlib_ext-stamp/qatlib_ext-build-.cmake` ...

in a vstart env there is a `*-RelWithDebInfo` there like in the below examples:
```
find . -name '*.cmake' | grep qatlib
./build/qatlib_ext-prefix/src/qatlib_ext-stamp/qatlib_ext-configure-RelWithDebInfo-impl.cmake
./build/qatlib_ext-prefix/src/qatlib_ext-stamp/qatlib_ext-configure-RelWithDebInfo.cmake
./build/qatlib_ext-prefix/src/qatlib_ext-stamp/qatlib_ext-build-RelWithDebInfo.cmake
./build/qatlib_ext-prefix/src/qatlib_ext-stamp/qatlib_ext-install-RelWithDebInfo-impl.cmake
./build/qatlib_ext-prefix/src/qatlib_ext-stamp/qatlib_ext-install-RelWithDebInfo.cmake
```

trying to debug with local rpmbuild

tchaikov reviewed (approved):
the CMake changes lgtm.

cbodley commented:
> all of the [shaman builds](https://shaman.ceph.com/builds/ceph/wip-cmake-qat/96d1aa1b8982a07a1b3509d2ab542783298e510e/) are struggling with qatlib:

thanks @mkogan1, i'd managed to work around those failures - the latest set of builds in https://shaman.ceph.com/builds/ceph/wip-cmake-qat/c7f3e29684b3817d8ea0342b7298045be05848e8/ show some green

tchaikov commented:
> > all of the [shaman builds](https://shaman.ceph.com/builds/ceph/wip-cmake-qat/96d1aa1b8982a07a1b3509d2ab542783298e510e/) are struggling with qatlib:
> > ```
> > [  7%] Performing build step for 'qatlib_ext'
> > CMake Error at /home/jenkins-build/build/workspace/ceph-dev-new-build/ARCH/x86_64/AVAILABLE_ARCH/x86_64/AVAILABLE_DIST/centos8/DIST/centos8/MACHINE_SIZE/gigantic/release/19.0.0-45-g96d1aa1b/rpm/el8/BUILD/ceph-19.0.0-45-g96d1aa1b/x86_64-redhat-linux-gnu/qatlib_ext-prefix/src/qatlib_ext-stamp/qatlib_ext-build-.cmake:37 (message):
> >   Command failed: 1
> > ```
> 
> noticed that there isent anything between the `qatlib_ext-build-` and the `.cmake` like in the ... `qatlib_ext-prefix/src/qatlib_ext-stamp/qatlib_ext-build-.cmake` ...
> 
> in a vstart env there is a `*-RelWithDebInfo` there like in the below examples:
> 
> ```
> find . -name '*.cmake' | grep qatlib
> ./build/qatlib_ext-prefix/src/qatlib_ext-stamp/qatlib_ext-configure-RelWithDebInfo-impl.cmake
> ./build/qatlib_ext-prefix/src/qatlib_ext-stamp/qatlib_ext-configure-RelWithDebInfo.cmake
> ./build/qatlib_ext-prefix/src/qatlib_ext-stamp/qatlib_ext-build-RelWithDebInfo.cmake
> ./build/qatlib_ext-prefix/src/qatlib_ext-stamp/qatlib_ext-install-RelWithDebInfo-impl.cmake
> ./build/qatlib_ext-prefix/src/qatlib_ext-stamp/qatlib_ext-install-RelWithDebInfo.cmake
> ```
> 
> trying to debug with local rpmbuild

@mkogan1 it's expected. because when building the packages, `CMAKE_BUILD_TYPE` is empty, and CMake encodes `CMAKE_BUILD_TYPE` into the paths for. 

i am looking at https://jenkins.ceph.com/job/ceph-dev-new-build/ARCH=x86_64,AVAILABLE_ARCH=x86_64,AVAILABLE_DIST=centos8,DIST=centos8,MACHINE_SIZE=gigantic/75581//consoleFull, where we have
```
[ 51%] Building CXX object src/librbd/CMakeFiles/rbd_internal.dir/api/DiffIterate.cc.o
In file included from /home/jenkins-build/build/workspace/ceph-dev-new-build/ARCH/x86_64/AVAILABLE_ARCH/x86_64/AVAILABLE_DIST/centos8/DIST/centos8/MACHINE_SIZE/gigantic/release/19.0.0-562-gc7f3e296/rpm/el8/BUILD/ceph-19.0.0-562-gc7f3e296/src/common/ceph_context.h:34,
                 from /home/jenkins-build/build/workspace/ceph-dev-new-build/ARCH/x86_64/AVAILABLE_ARCH/x86_64/AVAILABLE_DIST/centos8/DIST/centos8/MACHINE_SIZE/gigantic/release/19.0.0-562-gc7f3e296/rpm/el8/BUILD/ceph-19.0.0-562-gc7f3e296/src/common/dout.h:29,
                 from /home/jenkins-build/build/workspace/ceph-dev-new-build/ARCH/x86_64/AVAILABLE_ARCH/x86_64/AVAILABLE_DIST/centos8/DIST/centos8/MACHINE_SIZE/gigantic/release/19.0.0-562-gc7f3e296/rpm/el8/BUILD/ceph-19.0.0-562-gc7f3e296/src/include/Context.h:19,
                 from /home/jenkins-build/build/workspace/ceph-dev-new-build/ARCH/x86_64/AVAILABLE_ARCH/x86_64/AVAILABLE_DIST/centos8/DIST/centos8/MACHINE_SIZE/gigantic/release/19.0.0-562-gc7f3e296/rpm/el8/BUILD/ceph-19.0.0-562-gc7f3e296/src/common/Readahead.h:10,
                 from /home/jenkins-build/build/workspace/ceph-dev-new-build/ARCH/x86_64/AVAILABLE_ARCH/x86_64/AVAILABLE_DIST/centos8/DIST/centos8/MACHINE_SIZE/gigantic/release/19.0.0-562-gc7f3e296/rpm/el8/BUILD/ceph-19.0.0-562-gc7f3e296/src/librbd/ImageCtx.h:20,
                 from /home/jenkins-build/build/workspace/ceph-dev-new-build/ARCH/x86_64/AVAILABLE_ARCH/x86_64/AVAILABLE_DIST/centos8/DIST/centos8/MACHINE_SIZE/gigantic/release/19.0.0-562-gc7f3e296/rpm/el8/BUILD/ceph-19.0.0-562-gc7f3e296/src/librbd/api/DiffIterate.cc:5:
/home/jenkins-build/build/workspace/ceph-dev-new-build/ARCH/x86_64/AVAILABLE_ARCH/x86_64/AVAILABLE_DIST/centos8/DIST/centos8/MACHINE_SIZE/gigantic/release/19.0.0-562-gc7f3e296/rpm/el8/BUILD/ceph-19.0.0-562-gc7f3e296/src/common/cmdparse.h:14:10: fatal error: common/BackTrace.h: No such file or directory
```

this is indeed weird..

mkogan1 commented:
> https://jenkins.ceph.com/job/ceph-dev-new-build/ARCH=x86_64,AVAILABLE_ARCH=x86_64,AVAILABLE_DIST=centos8,DIST=centos8,MACHINE_SIZE=gigantic/75581//consoleFull,
...
> /home/jenkins-build/build/workspace/ceph-dev-new-build/ARCH/x86_64/AVAILABLE_ARCH/x86_64/AVAILABLE_DIST/centos8/DIST/centos8/MACHINE_SIZE/gigantic/release/19.0.0-562-gc7f3e296/rpm/el8/BUILD/ceph-19.0.0-562-gc7f3e296/src/common/cmdparse.h:14:10: fatal error: common/BackTrace.h: No such file or directory

yes, its not the only missing file there, myb tthat machines needs fsck ...

checking the locate rpmbuild qatlib_ext-build.log log, the only issues are several of those:
```
less /home/mkogan/rpmbuild/BUILD/ceph-19.0.0-592-gf5a281c22ac/x86_64-redhat-linux-gnu/qatlib_ext-prefix/src/qatlib_ext-stamp/qatlib_ext-build.log
quickassist/lookaside/access_layer/src/sample_code/performance/framework/linux/user_space/cpa_sample_code_utils.c: In function ‘getCompressedFile’:                                         
quickassist/lookaside/access_layer/src/sample_code/performance/framework/linux/user_space/cpa_sample_code_utils.c:987:47: warning: ‘%s’ directive output truncated writing 110 bytes into a >
  987 |         snprintf(fullpath, sizeof(fullpath), "%s%s", firmwarePath, filename);
      |                                               ^~
In file included from /usr/include/stdio.h:873,
                 from ./quickassist/include/cpa_types.h:97,
                 from ./quickassist/include/cpa.h:100,
                 from quickassist/lookaside/access_layer/src/sample_code/performance/framework/linux/user_space/cpa_sample_code_utils.h:74,                                                 
                 from quickassist/lookaside/access_layer/src/sample_code/performance/framework/linux/user_space/cpa_sample_code_utils.c:72:                                                 
/usr/include/bits/stdio2.h:68:10: note: ‘__builtin___snprintf_chk’ output 111 or more bytes into a destination of size 100                                                                  
   68 |   return __builtin___snprintf_chk (__s, __n, __USE_FORTIFY_LEVEL - 1,
      |          ^~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
   69 |                                    __glibc_objsize (__s), __fmt,
      |                                    ~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
   70 |                                    __va_arg_pack ());
      |                                    ~~~~~~~~~~~~~~~~~
```
from various samples, myb samples compilation can be disabled
(although on my machine those are only warnings depending on build machine gcc ver/hardening setup myb its the issue)

tchaikov commented:
ahh, infra issues .. that explains. thanks!

mkogan1 commented on cmake/modules/BuildQAT.cmake:
```diff
@@ -0,0 +1,48 @@
+  list(APPEND configure_cmd --disable-systemd)
+  # build a static library with -fPIC that we can link into crypto/compressor plugins
+  list(APPEND configure_cmd --with-pic --enable-static --disable-shared)
+
```
@cbodley suggesting
```suggestion
# disable the building of samples
list(APPEND configure_cmd --enable-samples=no)

```
tested to work locally
```
less ./qatlib_ext-prefix/src/qatlib_ext-stamp/qatlib_ext-configure.log
...
checking force disable fast crc in assembler... no
checking whether to build samples... no
                                     ^^
checking that generated files are newer than configure... done
...
```

cbodley replied:
what was the problem with samples? the ci builds in https://shaman.ceph.com/builds/ceph/wip-cmake-qat/bf90dbbbfb6ec1b998369749506e590d3d9dbac6/ seem ok there

liumiaomiaoIntel commented on .gitmodules:
```diff
@@ -75,3 +75,9 @@
 [submodule "src/jaegertracing/opentelemetry-cpp"]
 	path = src/jaegertracing/opentelemetry-cpp
 	url = https://github.com/open-telemetry/opentelemetry-cpp.git
+[submodule "src/qatlib"]
```
IMHO, it's not be appropriate to always download the latest version of the qatlib and QATzip repositories. Maybe branch [23.02.0](https://github.com/intel/qatlib/tree/23.02.0) for qatlib, branch [v1.1.2](https://github.com/intel/QATzip/tree/v1.1.2) for QATzip.

cbodley replied:
the submodules point to refs https://github.com/intel/qatlib/tree/142e305970ec66a860945d20bb7c330f99ed900b and https://github.com/intel/qatzip/tree/fdee557b5bb640827758f121102dcf3583292b7a, which correspond to the latest qatlib 23.11.0 and qatzip 1.1.2 releases

liumiaomiaoIntel replied:
got it. Learned how to use git submodule to add submodules

cbodley commented:
the rgw suite passed without regressions in https://pulpito.ceph.com/cbodley-2024-01-08_21:17:33-rgw-wip-cmake-qat-distro-default-smithi/

cbodley commented on src/crypto/CMakeLists.txt:
```diff
@@ -7,6 +7,6 @@ if(HAVE_INTEL AND HAVE_NASM_X64_AVX2 AND (NOT APPLE))
 endif()
 
-if(WITH_QAT)
+if(HAVE_QAT)
```
HAVE_QAT is true if either version (WITH_QATDRV or WITH_QATLIB) is set. isn't that what we want?

hualongfeng replied:
Yes, this is what we want. 

I want to solve ``../src/crypto/qat/qcccrypto.h:39:10: fatal error: qat/qae_mem_utils.h: No such file or directory`.

We can remove `#include <qat/qae_mem_utils.h>`  in `HAVE_QAT` in qcccrypto.h and add `#ifdef HAVE_QATDRV` for `qaeMemInit()` and `qaeMemDestroy()`.
```
//src/crypto/qat/qcccrypto.h
#ifdef HAVE_QATDRV
#include "cpa.h"
#include "cpa_cy_sym_dp.h"
#include "cpa_cy_im.h"
#include "lac/cpa_cy_sym.h"
#include "lac/cpa_cy_im.h"
#include "qae_mem.h"
#include "icp_sal_user.h"
#include "icp_sal_poll.h"
#include "qae_mem_utils.h"
#else
#include <qat/cpa.h>
#include <qat/cpa_cy_im.h>
#include <qat/cpa_cy_sym_dp.h>
#include <qat/cpa_cy_sym.h>
#include <qat/qae_mem.h>
#include <qat/icp_sal_user.h>
#include <qat/icp_sal_poll.h>
#endif


//src/crypto/qat/qcccrypto.cc
#ifdef HAVE_QATDRV
  stat = qaeMemInit();
  if (stat != CPA_STATUS_SUCCESS) {
    derr << "Unable to load memory driver" << dendl;
    this->cleanup();
    return false;
  }
#endif

#ifdef HAVE_QATDRV
  qaeMemDestroy();
#endif
```

cbodley replied:
thanks, that helps a lot:
> Ceph includes a qae_mem_utils.h header in [src/crypto/qat/qcccrypto.h](https://github.com/ceph/ceph/blob/07614d51c26cb3e98c87f4d9984a033b1e11761d/src/crypto/qat/qcccrypto.h#L30) that isn't present in the qatlib-devel package or the 'make install' directory of a qatlib build from source. i've opened https://github.com/intel/qatlib/pull/64 to address this upstream, and added a workaround to BuildQAT.cmake that manually copies the header into its install dir. until this is resolved and backported, we won't be able to rely on the qatlib-devel package

cbodley commented:
pushed changes:
* rebased
* added commit 'crypto/qat: qualify spawn::yield_context' after https://github.com/ceph/ceph/pull/50055
* made qae_mem_utils.h include conditional on HAVE_QATDRV
* removed workaround to copy qae_mem_utils.h in BuildQAT.cmake
* enabled WITH_SYSTEM_QATLIB by default for centos9

cbodley commented:
jenkins test this please

cbodley commented:
jenkins test this please

cbodley commented:
> The following tests FAILED:
	240 - unittest-seastar-socket (Subprocess aborted)

cbodley commented:
jenkins test make check

cbodley commented:
the latest builds are all green in https://shaman.ceph.com/builds/ceph/wip-cmake-qat/84af8357d1505c1b52cada7e4cdd4ad6e79e2382/

it's passed rgw qa and has an approval. i think it's ready to merge once the pr checks pass

cbodley commented:
> The following tests FAILED:
	240 - unittest-seastar-socket (Subprocess aborted)

cbodley commented:
jenkins test make check
